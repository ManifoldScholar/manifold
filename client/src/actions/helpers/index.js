export { default as createApiAction } from "./createApiAction";
export { default as commonActions } from "./commonActions";
