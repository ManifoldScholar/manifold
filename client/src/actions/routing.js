import { createAction } from "redux-actions";

export const update = createAction("ROUTE_UPDATE", state => state);
