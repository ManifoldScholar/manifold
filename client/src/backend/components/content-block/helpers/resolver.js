import blocks from "../Builder/Block/types";
import forms from "../TypeForm/types";
import contentBlockHelpers from "helpers/contentBlockHelpers";
import transform from "lodash/transform";

class Resolver {
  blockComponentsByType() {
    /* eslint-disable no-param-reassign */
    return transform(blocks, (result, val, key) => {
      result[contentBlockHelpers.componentKeyToType(key)] = val;
    });
    /* eslint-enable no-param-reassign */
  }

  typeToBlockComponent(type) {
    return contentBlockHelpers.typeToComponent(type, blocks);
  }

  typeToFormComponent(type) {
    return contentBlockHelpers.typeToComponent(type, forms);
  }
}

export default new Resolver();
