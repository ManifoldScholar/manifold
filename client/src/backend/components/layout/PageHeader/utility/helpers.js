import lh from "helpers/linkHandler";

export const getTextLinks = ({ texts, pathname }) =>
  texts?.length
    ? texts?.map(t => ({
        label: t.label,
        route: "backendTextAnalytics",
        id: t.id,
        active: pathname?.includes(t.id)
      }))
    : null;

export const getIssueLinks = ({ issues, pathname }) =>
  issues?.length
    ? issues?.map(i => ({
        label: i.label,
        route: "backendProjectAnalytics",
        id: i.id,
        active: pathname?.includes(i.id)
      }))
    : null;

export const getLinkOrButtonProps = action => {
  if (action.href) {
    return action.download
      ? { as: "a", href: action.href, download: true }
      : {
          as: "a",
          href: action.href,
          target: "_blank",
          re: "noopener noreferrer"
        };
  }

  if (action.route)
    return {
      to: lh.link(action.route, action.slug, action.resourceSlug)
    };

  return { as: "button", onClick: action.onClick };
};
