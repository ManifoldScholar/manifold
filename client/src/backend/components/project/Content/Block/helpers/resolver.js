import types from "../types";

class Resolver {
  typeToComponentKey(type) {
    const matches = type.match(/^Content::(.*)Block$/);
    if (!matches) return null;
    return matches[1];
  }

  typeToComponent(type) {
    const key = this.typeToComponentKey(type);
    if (!key) return null;
    const component = types[key];
    return component || null;
  }

  typeToAvailableId(type) {
    return type;
  }

  typeToCurrentId(id) {
    return `current-${id}`;
  }
}

export default new Resolver();
