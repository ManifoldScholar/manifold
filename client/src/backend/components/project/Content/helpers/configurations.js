class Configurations {
  get configurationMap() {
    return {
      "Content::HeroBlock": true,
      "Content::RecentActivityBlock": false,
      "Content::TextsBlock": true,
      "Content::ResourcesBlock": true,
      "Content::MetadataBlock": false,
      "Content::TableOfContentsBlock": true,
      "Content::MarkdownBlock": true
    };
  }

  isConfigurable(type) {
    return this.configurationMap[type];
  }
}

export default new Configurations();
