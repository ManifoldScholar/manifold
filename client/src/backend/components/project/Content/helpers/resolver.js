import blocks from "../Block/types";
import forms from "../TypeForm/types";

class Resolver {
  typeToComponentKey(type) {
    const matches = type.match(/^Content::(.*)Block$/);
    if (!matches) return null;
    return matches[1];
  }

  typeToComponent(type, types) {
    const key = this.typeToComponentKey(type);
    if (!key) return null;
    const component = types[key];
    return component || null;
  }

  typeToBlockComponent(type) {
    return this.typeToComponent(type, blocks);
  }

  typeToFormComponent(type) {
    return this.typeToComponent(type, forms);
  }

  typeToAvailableId(type) {
    return type;
  }

  typeToCurrentId(id) {
    return `current-${id}`;
  }
}

export default new Resolver();
