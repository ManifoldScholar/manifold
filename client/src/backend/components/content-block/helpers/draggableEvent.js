import startsWith from "lodash/startsWith";
import resolver from "./resolver";
import sumBy from "lodash/sumBy";
import cloneDeep from "lodash/cloneDeep";

export default class DraggableEventHelper {
  static syntheticDraggable = (blockType, opts = {}) => {
    const options = {
      ...DraggableEventHelper.defaultOptions(),
      ...opts
    };
    return {
      type: DraggableEventHelper.isTopType(blockType) ? "TOP" : "BOTTOM",
      draggableId: blockType,
      source: {
        droppableId: options.source.name
      },
      destination: {
        droppableId: options.destination.name,
        index: -1
      }
    };
  };

  static isTopType = type => {
    return !!resolver.typeToBlockComponent(type).top;
  };

  static defaultOptions = () => {
    return {
      source: {
        name: "available"
      },
      destination: {
        name: "current"
      }
    };
  };

  constructor(draggable, blocks, options = {}) {
    this.options = { ...this.defaultOptions, ...options };
    this.draggable = draggable;
    this._blocks = blocks;
  }

  get block() {
    if (this.isMove) return this.movedBlock;
    if (this.isInsert) return this.insertedBlock;
  }

  get blocks() {
    let blocks = this.clonedBlocks;
    const block = this.block;
    if (this.isMove) {
      blocks = blocks.filter(b => b.id !== block.id);
    }
    blocks.splice(block.attributes.position - 1, 0, block);
    return blocks;
  }

  get defaultOptions() {
    return this.constructor.defaultOptions();
  }

  get clonedBlocks() {
    return Array.from(this._blocks);
  }

  get draggedFromSource() {
    if (!this.draggable.source) return false;
    return startsWith(
      this.draggable.source.droppableId,
      this.options.source.name
    );
  }

  get draggedFromDestination() {
    if (!this.draggable.source) return false;
    return startsWith(
      this.draggable.source.droppableId,
      this.options.destination.name
    );
  }

  get draggedToDestination() {
    if (!this.draggable.destination) return false;
    return startsWith(
      this.draggable.destination.droppableId,
      this.options.destination.name
    );
  }

  get destinationIndex() {
    if (!this.draggable.destination) return null;
    const index = this.draggable.destination.index;
    if (index === -1) return this.autoIndex;
    return index;
  }

  get autoIndex() {
    if (this.isTopType(this.type)) return this.topCount; // no +1 because index is from 0;
    return this._blocks.length;
  }

  get actionable() {
    return this.action != null;
  }

  get isMove() {
    return this.draggedFromDestination && this.draggedToDestination;
  }

  get isInsert() {
    return this.draggedFromSource && this.draggedToDestination;
  }

  get action() {
    if (this.isMove) return "move";
    if (this.isInsert) return "insert";
    return null;
  }

  get isTop() {
    return this.draggable.type === "TOP";
  }

  isTopBlock(block) {
    return this.isTopType(block.attributes.type);
  }

  isTopType(type) {
    return this.constructor.isTopType(type);
  }

  get topCount() {
    return sumBy(this._blocks, block => (this.isTopBlock(block) ? 1 : 0));
  }

  get type() {
    return this.draggable.draggableId;
  }

  get uuid() {
    return this.draggable.draggableId;
  }

  get position() {
    const index = this.destinationIndex;
    const offset = 1; // position starts from 1, index from 0
    if (this.isTop) return index + offset;
    // Blocks in the bottom list have to add the top count.
    return index + this.topCount + offset;
  }

  findBlock(id) {
    return this._blocks.find(block => block.id === id);
  }

  cloneBlock(id) {
    return cloneDeep(this.findBlock(id));
  }

  get insertedBlock() {
    return {
      id: "pending",
      attributes: {
        type: this.type,
        position: this.position,
        renderable: true
      },
      relationships: {}
    };
  }

  get movedBlock() {
    const block = this.cloneBlock(this.uuid);
    block.attributes.position = this.position;
    return block;
  }
}
