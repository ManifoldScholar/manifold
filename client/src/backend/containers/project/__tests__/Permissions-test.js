import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import ProjectPermissionsContainer from "../Permissions";
import { wrapWithRouter } from "test/helpers/routing";
import build from "test/fixtures/build";

describe("Backend Project Permissions Container", () => {
  const project = build.entity.project("1");
  const route = {
    routes: [],
    options: {}
  };

  const root = <ProjectPermissionsContainer project={project} route={route} />;

  it("renders correctly", () => {
    const renderer = new ShallowRenderer();
    const tree = renderer.render(root);
    expect(tree).toMatchSnapshot();
  });

  it("doesn't render to null", () => {
    const renderer = new ShallowRenderer();
    const tree = renderer.render(root);
    expect(tree).not.toBe(null);
  });
});
