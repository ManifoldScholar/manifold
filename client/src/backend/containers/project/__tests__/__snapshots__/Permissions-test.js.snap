// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backend Project Permissions Container renders correctly 1`] = `
<Connect(withRouter(AuthorizeComponent))
  ability="managePermissions"
  entity={
    Object {
      "attributes": Object {
        "abilities": Object {
          "create": true,
          "createPermissions": true,
          "createResourceCollections": true,
          "createResources": true,
          "createTexts": true,
          "createTwitterQueries": true,
          "delete": true,
          "manageEvents": true,
          "managePermissions": true,
          "manageResourceCollections": true,
          "manageResources": true,
          "manageSocials": true,
          "manageTexts": true,
          "manageTwitterQueries": true,
          "read": true,
          "readDeleted": true,
          "readDrafts": true,
          "readLog": true,
          "readSecrets": true,
          "update": true,
          "updateMakers": true,
        },
        "avatarColor": "primary",
        "avatarMeta": Object {},
        "avatarStyles": Object {},
        "coverStyles": Object {},
        "darkMode": false,
        "description": "The look in your eyes, I recognize it. You used to have it for me. I guess it's better to be lucky than good. Mr. Crusher, ready a collision course with the Borg ship. My oath is between Captain Kargan and myself. Your only concern is with how you obey my orders. Or do you prefer the rank of prisoner to that of lieutenant? That might've been one of the shortest assignments in the history of Starfleet. I'll alert the crew.",
        "descriptionFormatted": "The look in your eyes, I recognize it. You used to have it for me. I guess it's better to be lucky than good. Mr. Crusher, ready a collision course with the Borg ship. My oath is between Captain Kargan and myself. Your only concern is with how you obey my orders. Or do you prefer the rank of prisoner to that of lieutenant? That might've been one of the shortest assignments in the history of Starfleet. I'll alert the crew.",
        "downloadCallToAction": "Download the greatest dog",
        "hashtag": "cute_dog",
        "heroStyles": Object {},
        "hideActivity": false,
        "imageCreditsFormatted": "<p>Rowan, <em>The Dog</em></p>",
        "metadata": Object {
          "doi": "https://doi.org/10.12345.6789",
        },
        "metadataFormatted": Object {
          "doi": "https://doi.org/10.12345.6789",
        },
        "metadataProperties": Array [
          "abstract",
          "archive",
          "archiveLocation",
          "archivePlace",
          "authority",
          "callNumber",
          "collectionTitle",
          "containerTitle",
          "dimensions",
          "event",
          "eventPlace",
          "isbn",
          "issn",
          "jurisdiction",
          "medium",
          "originalPublisher",
          "originalPublisherPlace",
          "originalTitle",
          "pmcid",
          "pmid",
          "publisher",
          "publisherPlace",
          "reviewedTitle",
          "section",
          "version",
          "yearSuffix",
          "chapterNumber",
          "collectionNumber",
          "edition",
          "issue",
          "number",
          "numberOfPages",
          "numberOfVolumes",
          "volume",
        ],
        "slug": "slug-1",
        "subtitle": "World's Greatest Dog",
        "title": "Rowan Test",
        "titleFormatted": "Rowan Test",
        "titlePlaintext": "Rowan Test",
      },
      "id": "1",
      "relationships": Object {
        "actionCallouts": Array [],
        "makers": Array [
          Object {
            "attributes": Object {
              "firstName": "John",
              "fullName": "John Milton",
              "lastName": "Milton",
            },
            "type": "maker",
          },
        ],
        "resources": Array [],
      },
      "type": "projects",
    }
  }
  failureNotification={true}
  failureRedirect="/backend/projects/1"
>
  <section>
    <withRouter(Connect(HigherOrder.FetchData('PermissionContainer)))
      entity={
        Object {
          "attributes": Object {
            "abilities": Object {
              "create": true,
              "createPermissions": true,
              "createResourceCollections": true,
              "createResources": true,
              "createTexts": true,
              "createTwitterQueries": true,
              "delete": true,
              "manageEvents": true,
              "managePermissions": true,
              "manageResourceCollections": true,
              "manageResources": true,
              "manageSocials": true,
              "manageTexts": true,
              "manageTwitterQueries": true,
              "read": true,
              "readDeleted": true,
              "readDrafts": true,
              "readLog": true,
              "readSecrets": true,
              "update": true,
              "updateMakers": true,
            },
            "avatarColor": "primary",
            "avatarMeta": Object {},
            "avatarStyles": Object {},
            "coverStyles": Object {},
            "darkMode": false,
            "description": "The look in your eyes, I recognize it. You used to have it for me. I guess it's better to be lucky than good. Mr. Crusher, ready a collision course with the Borg ship. My oath is between Captain Kargan and myself. Your only concern is with how you obey my orders. Or do you prefer the rank of prisoner to that of lieutenant? That might've been one of the shortest assignments in the history of Starfleet. I'll alert the crew.",
            "descriptionFormatted": "The look in your eyes, I recognize it. You used to have it for me. I guess it's better to be lucky than good. Mr. Crusher, ready a collision course with the Borg ship. My oath is between Captain Kargan and myself. Your only concern is with how you obey my orders. Or do you prefer the rank of prisoner to that of lieutenant? That might've been one of the shortest assignments in the history of Starfleet. I'll alert the crew.",
            "downloadCallToAction": "Download the greatest dog",
            "hashtag": "cute_dog",
            "heroStyles": Object {},
            "hideActivity": false,
            "imageCreditsFormatted": "<p>Rowan, <em>The Dog</em></p>",
            "metadata": Object {
              "doi": "https://doi.org/10.12345.6789",
            },
            "metadataFormatted": Object {
              "doi": "https://doi.org/10.12345.6789",
            },
            "metadataProperties": Array [
              "abstract",
              "archive",
              "archiveLocation",
              "archivePlace",
              "authority",
              "callNumber",
              "collectionTitle",
              "containerTitle",
              "dimensions",
              "event",
              "eventPlace",
              "isbn",
              "issn",
              "jurisdiction",
              "medium",
              "originalPublisher",
              "originalPublisherPlace",
              "originalTitle",
              "pmcid",
              "pmid",
              "publisher",
              "publisherPlace",
              "reviewedTitle",
              "section",
              "version",
              "yearSuffix",
              "chapterNumber",
              "collectionNumber",
              "edition",
              "issue",
              "number",
              "numberOfPages",
              "numberOfVolumes",
              "volume",
            ],
            "slug": "slug-1",
            "subtitle": "World's Greatest Dog",
            "title": "Rowan Test",
            "titleFormatted": "Rowan Test",
            "titlePlaintext": "Rowan Test",
          },
          "id": "1",
          "relationships": Object {
            "actionCallouts": Array [],
            "makers": Array [
              Object {
                "attributes": Object {
                  "firstName": "John",
                  "fullName": "John Milton",
                  "lastName": "Milton",
                },
                "type": "maker",
              },
            ],
            "resources": Array [],
          },
          "type": "projects",
        }
      }
    />
    <withRouter(Router.Switch('DrawerSwitch))
      closeUrl="/backend/projects/1/permissions"
      parentRoute={
        Object {
          "options": Object {},
          "routes": Array [],
        }
      }
    />
  </section>
</Connect(withRouter(AuthorizeComponent))>
`;
